Add codes for path types

jonsterling opened the pull request:
```
A : I -> U
a : (i : I) -> (p : [i=0\/i=1]) -> el(A i)
-------------
path(A, a) : U
el(path(A,a)) = (i : I) -> el(A i) [i=0\/i=1 => a i]
```


Resolves #34, #35, #36, #37 .

favonia commented:
Would this be consumed by #23?

favonia commented:
Would this be subsumed by #23?

jonsterling commented:
Yes; path types are a stepping stone, and we’re implementing them in a way that will generalize directly.

I want to hold off on extension types because I want a better story for ensuring that the cofibration is “local”. (Matt and I have been thinking of some lightweight box modality)



On Fri, Mar 20, 2020, at 6:35 PM, favonia wrote:
> 
> 
> Would this be subsumed by #23 <https://github.com/jonsterling/cooltt/issues/23>?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/jonsterling/cooltt/pull/33#issuecomment-601939306>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAANOEK3XKQWRF2S4Y322JDRIPVU5ANCNFSM4LQVR22A>.
>

jonsterling commented on src/lib/Nbe.ml:
```diff
@@ -814,8 +814,11 @@ struct
-      raise Todo
+    | univ, D.CodePath (fam, bdry) -> (* check *)
+      let* tfam = quote_con univ fam in
+      let+ tbdry = quote_con univ bdry
```
Very good guess! The type in which you are quoting the boundary is going to be:

```
    (i : I) -> [i=0\/i=1] -> fam i
```

We can work together to see how to translate that type into code.

ivoysey replied:
ah, ok so that's instead of the universe in the second recursive call?

jonsterling replied:
Yes, that's right

ivoysey replied:
this is now completed based on our session this morning, and with term builder instead of the explicit closures, so i'll "resolve this conversation".

jonsterling commented:
I have pushed a test file to guide the implementation of the rest of the path type; there are currently some bugs elsewhere that may prevent it from ultimately working, but this won't be noticed until a later stage of this mini-project, and I'm fixing them in the meanwhile.

jonsterling commented:
Update:

I have added two test files: one in `dream-tests/`, which travis won't execute by default, that contains examples that (if they worked) would (roughly) mean that we have successfully added path types to cooltt.

Second, I have added an analogous (passing) test to `tests/` that implements those examples in terms of the dependent product over the interval and the cubical subtype primitives; I did this to verify that it would be possible to carry out the task without getting owned by bugs that are my fault 😆  In fact, when creating this test, I did get owned by bugs that are my fault, but as you can see I have pushed fixes for them.

cangiuli commented on test/selfification.cooltt:
```diff
@@ -1,9 +1,8 @@
+  (z : Z) -> (x : sub A {0==0} {fst {p z}}) * B x
 } = {
-  \A B x => x
+  \_ _ _  p => p
```
:man_dancing:

jonsterling replied:
lol

jonsterling commented on src/lib/SyntaxData.ml:
```diff
@@ -45,5 +48,6 @@ and tp =
   | Sg of tp * tp
   | Id of tp * t * t
   | Nat
+  | Path of tp * t * t (*todo/iev: not quite sure i really understand why this should be in tp not t, given the relative flatness of coresyntaxdata.t*)
```
This line is unnecessary; the existing path type code is just going to be equipped with an equation like `El(path(A,a)) == (i : dim) -> [blah blah blah]`; this means, for instance, that there is also no need for `PathAp` or `PathLam`, nor is there a need for concrete syntax forms for path lambda/path application, since they will come already from the function type.

ivoysey replied:
per discussion the other day: got it

jonsterling commented on src/lib/Refiner.ml:
```diff
@@ -353,6 +353,41 @@ struct
+    ------------------------------------
+    path x.A a b type
+   *)
+  let formation (tac_tp : T.tp_tac) (tac0 : T.chk_tac) (tac1 : T.chk_tac) : T.tp_tac =
```
This rule will be unnecessary, because the path type is totally defined by the equation `el(path(...))`; so what we will need instead is a typing rule for the path _code_, something of type `chk_tac -> chk_tac -> chk_tac` or something along those lines.

ivoysey replied:
per discussion the other day: got it

jonsterling commented on src/lib/Refiner.ml:
```diff
@@ -353,6 +353,41 @@ struct
+    -------------------------------------
+    <x>M : path (x.A) a b
+  *)
+  let intro = raise Todo2
```
The intro and elim rule will actually be unnecessary, as they will be inherited automatically from the dependent product and cubical subtype connectives.

jonsterling commented on src/lib/Nbe.ml:
```diff
@@ -793,67 +814,81 @@ struct
+   *  path : (fam : I -> U) -> ((i : I) -> (p : [i=0\/i=1]) -> fam i) -> U
+   * *)
+    | univ, D.CodePath (fam, bdry) -> (* check *)
+      let* tfam = quote_con univ fam in
```
This line is incorrect, we will need to quote this thing at not `univ` but at `(i : dim) -> univ`. This type can be written down using the QQ/TB stuff (see the next line, `bdry_tp`, for an example of how to do this).

ivoysey commented:
got it; i'll clean that up and then you can merge.
